### ISSUE TITLE
Include stemming and lemmatization in the search execution

### ISSUE THREAD

**Issue by CharlieDigital** [typesense/typesense#834]
## Description

As an example, the search for a keyword "coding" will not return results with the word "code".

## Steps to reproduce

The following produces 43,816 results:

```
https://books-search.typesense.org/?b%5Bquery%5D=code

curl -sv 'https://s23uakwqftelv6idp.a1.typesense.net/multi_search?x-typesense-api-key=44aUZ6UOyg2tQlfNu89qVTTU2Kd9k1UL' \
  --data-raw '{"searches":[{"query_by":"author,title","sort_by":"","highlight_full_fields":"author,title","collection":"b","q":"code","facet_by":"author,subjects","max_facet_values":40,"page":1,"per_page":15}]}' \
  --compressed | grep "found"
```

The following produces 4,366 results:

```
https://books-search.typesense.org/?b%5Bquery%5D=coding

curl -sv 'https://s23uakwqftelv6idp.a1.typesense.net/multi_search?x-typesense-api-key=44aUZ6UOyg2tQlfNu89qVTTU2Kd9k1UL' \
  --data-raw '{"searches":[{"query_by":"author,title","sort_by":"","highlight_full_fields":"author,title","collection":"b","q":"coding","facet_by":"author,subjects","max_facet_values":40,"page":1,"per_page":15}]}' \
  --compressed | grep "found"
```

## Expected Behavior

The search for "coding" includes the results for the stemmed term "code" and has a total count of 43,816.

## Actual Behavior

The search for "coding" does not include the results for the stemmed term "code" and returns only 4,366 results.

In some use cases, it may be possible to use query parameters like `num_typo` an `infix` to achieve better results, but having the option to perform stemming on the search term(s) would improve the search experience for the end user.

Current workaround is to execute the stemming prior to submitting the search to Typesense.

## Metadata

**Typesense Version**: Live sample version

**OS**: Linux

**Comment by jonasjancarik** [comment]
> Current workaround is to execute the stemming prior to submitting the search to Typesense.

Does this mean Typesense can perform stemming or lemmatization at index time (similar to Elasticsearch when a language analyzer is used)?

**Comment by zopyx** [comment]
> > Current workaround is to execute the stemming prior to submitting the search to Typesense.
> 
> Does this mean Typesense can perform stemming or lemmatization at index time (similar to Elasticsearch when a language analyzer is used)?

In my understanding: NO. Stemming your own content as part of your import pipeline into Typesense should be straight forward. Since one of the key feature of Typesense is `find-as-you-type`, stemming is clearly less important (or even unimportant).

**Comment by kishorenc** [comment]
We're picking up this feature for development, so you can expect stemming support to land in the near future. I will update this thread.

**Comment by kishorenc** [comment]
This feature is available in `0.26.0.rc56`.

Usage:

```
{
  "name":"docs",
  "fields":[
    {
      "name":"title",
      "type":"string",
      "stem":true
    }
  ]
}
```

The title field will now be stemmed before it is indexed in-memory. The actual value stored on disk is not affected.

**Comment by zopyx** [comment]
> This feature is available in `0.26.0.rc56`.
> 
> Usage:
> 
> ```
> {
>   "name":"docs",
>   "fields":[
>     {
>       "name":"title",
>       "type":"string",
>       "stem":true
>     }
>   ]
> }
> ```
> 
> The title field will now be stemmed before it is indexed in-memory. The actual value stored on disk is not affected.

Stemming is language-specific. How do you indicate the language of a field?

**Comment by kishorenc** [comment]
The `locale` property of the field is used for language selection. I will also update the snippet above with this information.

**Comment by zopyx** [comment]
This might be a tricky if you have a field containing texts in different languages. We don't use stemming, but we have a `text` field that contains either text in `de` or `en` - by design. In such a case, the stemmer does not know the language to be used for stemming.

**Comment by tarjei** [comment]
Any timeframe for when this will land in the cloud edition?

**Comment by jasonbosco** [comment]
@tarjei `0.26.0.rc58` is already available in Typesense Cloud. You should find it at the bottom of the Typesense Version drop down.

**Comment by jasonbosco** [comment]
Released in v26.0
